### PR TITLE
llama.vim : add classic vim support

### PR DESCRIPTION
**Pull request by m18coppola**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

**Comment by ggerganov** [comment]
Testing with Vim 9.1, the first call to `prop_add` when displaying the ghost text causes the window to shift up a few lines. Happens when editing a file that is larger than the vertical screen size. Can you reproduce?

Also the async jobs seem to create many scratch buffers with the following contents:

image

**Comment by m18coppola** [comment]
@ggerganov I fixed the scratch buffer issue, it was due to a bad option for the `job_start()`. I was unable to reproduce the window shifting problem - I suspect that it may be because of the buffer-spam though. Retest and see if it fixes your issue. If the window shifting bug is still present after this fix, attach your vimrc and your `vim --version` output.

File: examples/llama.vim [hunk 1]
```diff
@@ -671,15 +733,61 @@ function! s:fim_on_stdout(job_id, data, event) dict
...
+            call prop_add(line('.'), col('.'), {
+                        \ 'type': s:hint_hlgroup,
+                        \ 'text': l:new_suffix . l:inset
+                        \ })
+        endif
+        for line in s:content[1:]
+            call prop_add(line('.'), 0, {
+                        \ 'type': s:hint_hlgroup,
+                        \ 'text': l:new_suffix . line,
+                        \ 'text_padding_left': s:get_indent(line),
```
Comment (ggerganov) [inline examples/llama.vim:778]: This padding does not seem to be necessary:

```suggestion
```

Comment (m18coppola) [reply examples/llama.vim:778]: snip57
For me, it seems to be required

Comment (ggerganov) [reply examples/llama.vim:778]: Hm, here it looks like this with the padding:

image

And without the padding:

image

Comment (m18coppola) [reply examples/llama.vim:778]: I think I'm forgetting to address the tabstop. What do you get as output when you execute `:echo &shiftwidth | :echo &tabstop`? What are your vimrc settings regarding tabbing/whitespace?

Comment (ggerganov) [reply examples/llama.vim:778]:
```
:echo &shiftwidth | :echo &tabstop
4
4
```

This is my [.vimrc](https://github.com/ggerganov/ggterm/blob/master/.vimrc)

Comment (m18coppola) [reply examples/llama.vim:776]: I think it's fixed now

File: examples/llama.vim [hunk 2]
```diff
@@ -93,6 +93,33 @@ let s:default_config = {
 
 let g:llama_config = get(g:, 'llama_config', s:default_config)
 
+let s:nvim_ghost_text = exists('*nvim_buf_get_mark')
+let s:vim_ghost_text = has('textprop')
+
+if s:vim_ghost_text
+    let s:hint_hlgroup = 'llama_hl_hint'
+    let s:info_hlgroup = 'llama_hl_info'
```
Comment (ggerganov) [inline examples/llama.vim:101]: This is more inline with the naming convention that we normally use in the project:

```suggestion
    let s:hlgroup_hint = 'llama_hl_hint'
    let s:hlgroup_info = 'llama_hl_info'
```

File: examples/llama.vim [hunk 3]
```diff
@@ -93,6 +93,33 @@ let s:default_config = {
...
+
+    if empty(prop_type_get(s:hint_hlgroup))
+        call prop_type_add(s:hint_hlgroup, {'highlight': s:hint_hlgroup})
+    endif
+    if empty(prop_type_get(s:info_hlgroup))
+        call prop_type_add(s:info_hlgroup, {'highlight': s:info_hlgroup})
+    endif
```
Comment (ggerganov) [inline examples/llama.vim]: These `s:` variables should go inside `llama#init` after `let s:current_job` and before `augroup llama` in order to have them all together in one place.

File: examples/llama.vim [hunk 4]
```diff
@@ -93,6 +93,33 @@ let s:default_config = {
 
 let g:llama_config = get(g:, 'llama_config', s:default_config)
 
+let s:nvim_ghost_text = exists('*nvim_buf_get_mark')
+let s:vim_ghost_text = has('textprop')
```
Comment (ggerganov) [inline examples/llama.vim:97]: nit:

```suggestion
let s:ghost_text_nvim = exists('*nvim_buf_get_mark')
let s:ghost_text_vim  = has('textprop')
```

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
@m18coppola Thanks for implementing the classic Vim support!

In case you are using this plugin, would appreciate any feedback about issues or ideas to improve it. So far, I think the biggest drawback compared to the original GH copilot is that it tends to over-suggest in certain situations (i.e. repeats lines that are in the suffix). The indentation logic helped a lot to limit this, but it seems that there is still something missing. Not sure if the sampler has to be improved in some way or if it is a model problem. Anyway, if you have any thoughts, let me know.

**Pull request merged**